R/data.table.R

```diff
@@ -2335,21 +2335,18 @@ copy = function(x) {
   newx = .Call(Ccopy,x)  # copies at length but R's duplicate() also copies truelength over.
                          # TO DO: inside Ccopy it could reset tl to 0 or length, but no matter as selfrefok detects it
                          # TO DO: revisit duplicate.c in R 3.0.3 and see where it's at
-  if (!is.data.table(x)) {
-    # fix for #1476. TODO: find if a cleaner fix is possible..
-    if (is.list(x)) {
-      anydt = vapply_1b(x, is.data.table, use.names=FALSE)
-      if (sum(anydt)) {
-        newx[anydt] = lapply(newx[anydt], function(x) {
-          .Call(C_unlock, x)
-          setalloccol(x)
-        })
-      }
+
+  reallocate = function(y) {
+    if (is.data.table(y)) {
+      .Call(C_unlock, y)
+      setalloccol(y)
+    } else if (is.list(y)) {
```

Comment:
I guess `is.recursive` would be more appropriate here? IIUC it would apply to anything for which the `lapply(y, reallocate)` part would go through

Reply:
Good logic, but `is.recursive` is, unlike its name, and actually more like `!is.atomic`. For example, `is.recursive(mean)` is true. Feels right to me to use a direct `is.list` to target what's in mind, and postpone an expansion to `is.recurisve` until we know we need to and have an example and test.

Reply:
good observation, I wish it was in the doc s!

Reply:
`?is.recursive` does include "Most types of objects are regarded as recursive. Exceptions are ..." but then I'm not really sure what it is useful for vs `!is.atomic`.

Reply:
just happened upon the source for this as well:

```
/* is.recursive */
	switch(TYPEOF(CAR(args))) {
	case VECSXP:
	case LISTSXP:
	case CLOSXP:
	case ENVSXP:
	case PROMSXP:
	case LANGSXP:
	case SPECIALSXP:
	case BUILTINSXP:
	case DOTSXP:
	case ANYSXP:
	case EXPRSXP:
	// Not recursive, as long as not subsettable (on the R level)
	// case EXTPTRSXP:
	// case BCODESXP:
	// case WEAKREFSXP:
	    LOGICAL0(ans)[0] = 1;
	    break;
	default:
	    LOGICAL0(ans)[0] = 0;
	    break;
	}
```